### PR TITLE
Modernize project infrastructure

### PR DESCRIPTION
**Pull request by mdomke**
* Use `hatch` as build tool and for environment handling
* Use `pytest` as test-runner
* Use PEP 621 compliant project metadata (`pyproject.toml`)
* Fix some linter errors and test failures along the way
* Use GitHub Actions for CI
* Remove support for deprecated Python versions (everything prior to 3.8)

## Future plans
* Switch to `ruff`/`hatch` as linter and formatter
* ~~Use `black` as formatter and enforce formatting with `precommit` rules~~
* Modernize the code with `pyupgrade` to remove relics from Python2 times e.g (`class Foo(object):`)

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,15 +1,14 @@
...
+RUN apt-get update \
+ && apt-get install -y make build-essential libssl-dev zlib1g-dev libbz2-dev libreadline-dev libsqlite3-dev wget curl llvm libncurses5-dev git netcat
 RUN curl -L https://raw.githubusercontent.com/yyuu/pyenv-installer/master/bin/pyenv-installer | bash
 
 ENV PYENV_ROOT /root/.pyenv
 ENV PATH $PYENV_ROOT/shims:$PYENV_ROOT/bin:$PATH
 
 RUN eval "$(pyenv init -)" && eval "$(pyenv virtualenv-init -)"
-RUN pyenv install 3.6.15
-RUN pyenv install 3.7.9
```
Comment (pcorpet) [inline Dockerfile:12]: Make sure to update the pyenv local line below

Comment (pcorpet) [inline Dockerfile:12]: Please consider adding 3.11 and 3.12 as well. And also probably a more recent pypy3

File: docker-compose.yml [hunk 2]
```diff
@@ -19,3 +19,6 @@ services:
     # The image chosen here should be kept in sync with the .travis.yml mongo version and with
     # the value in mongomock/__init__.py
     image: mongo:5.0.5
+    ports:
+      - 27017:27017
```
Comment (pcorpet) [inline docker-compose.yml]: This poses a security threat: you're opening a port on the host. If you want to check locally, I'd recommend using
```yml
ports:
  - localhost:27017:27017
```

File: LICENSE [hunk 3]
```diff
@@ -1,24 +1,15 @@
-Copyright (c) 2012, Rotem Yaari <vmalloc@gmail.com>
```
Comment (pcorpet) [inline LICENSE:1]: You'll need @vmalloc's approval on that one.

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+[project]
+name = "mongomock"
+dynamic = ["version"]
+description = "Fake pymongo stub for testing simple MongoDB-dependent code"
+readme = { file = "README.rst", content-type = "text/x-rst" }
+license = { file = "LICENSE" }
+authors = [
+    { name = "Rotem Yaari", email = "vmalloc@gmail.com" },
+    { name = "Martin Domke", email = "mail@martindomke.net" },
+]
```
Comment (pcorpet) [inline pyproject.toml:15]: Please add myself in there `{ name = "Pascal Corpet", email = "pascal@corpet.net" },`

File: tox.ini [hunk 5]
```diff
@@ -1,13 +1,11 @@
...
-passenv = CI TRAVIS TRAVIS_* TEST_MONGO_HOST MONGODB
+passenv = CI, TRAVIS, TRAVIS_*, TEST_MONGO_HOST, MONGODB
 setenv =
   EXECJS_RUNTIME = Node
 basepython =
-  py36: python3.6
-  py37: python3.7
   py38: python3.8
   py39: python3.9
   py310: python3.10
```
Comment (pcorpet) [inline tox.ini:11]: Please add python 3.11 and python 3.12 (or remove them from the other places).

**Review by pcorpet — CHANGES_REQUESTED** [review]
Thanks a lot for this big cleanup. In general I'm more in favor of having multiple small PRs. If you can, please send some fixes in the python code separately from the changes of BUILD and separately from the changes in CI.

If too much of a hastle, let's continue it here although the review my be longer.

File: LICENSE (hunk 3 above)
Comment (mdomke) [reply LICENSE:1]: Yes, you're right. I'll try to contact him.

File: docker-compose.yml (hunk 2 above)
Comment (mdomke) [reply docker-compose.yml]: `localhost` doesn't work, but I can use `127.0.0.1` instead. I'm not sure about the attack vector however

File: tox.ini (hunk 5 above)
Comment (mdomke) [reply tox.ini:11]: The point is to eventually replace `tox` with `hatch`, but I can certainly add the additional Python versions for now

**Comment by mdomke** [comment]
Hej @pcorpet. I agree that pull-requests should be kept related to a specific change. If it is okay with you, I would like to avoid going the extra mile and slice this pull-request into several smaller ones. What do you think?

File: LICENSE (hunk 3 above)
Comment (mdomke) [reply LICENSE:1]: I haven't heard back so far from @vmalloc. I will change the license back for the time being

**Comment by mdomke** [comment]
Also, @pcorpet, could you add the PyPI API token as a GitLab secret so that we can also switch to GitHub Actions for the deployment?

**Comment by vmalloc** [comment]
@mdomke @pcorpet I just noticed this in the pile of notifications... Can you please explain what is needed here?

**Comment by mdomke** [comment]
@vmalloc Originally I wanted to update the license to the [ISC license](https://choosealicense.com/licenses/isc/) and change the copyright owners to "Mongomock Contributors". 

> The ISC license is functionally equivalent to the [BSD 2-Clause](https://choosealicense.com/licenses/bsd-2-clause/) and [MIT](https://choosealicense.com/licenses/mit/) licenses, removing some language that is no longer necessary.

@pcorpet rightfully made me aware that for this kind of change, we would need your approval and that's why I was reaching out to you. In the meantime, I switched (commit: [f44fc3e](https://github.com/mongomock/mongomock/pull/876/commits/f44fc3e66f3319680e383b53daeff36e2683a302)) the license back to the original BSD 2-Clause style to not block this PR.

**Comment by vmalloc** [comment]
So I'm good with the license change to ISC as long as my name is still somewhere in the authors list

**Comment by mdomke** [comment]
It is and always will be 😄  

See: https://github.com/mongomock/mongomock/pull/876/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R11-R15

**Comment by mdomke** [comment]
@pcorpet I would like to proceed with this pull-request and if it is just only for being able to use GitHub Actions. After clarifying the license change with @vmalloc I also included the ISC license again.

**Pull request merged**